File: sdk/react-native/dazzle-react-native/src/agent/contextStore.ts
// Copyright 2026 Ivan Aliaga <ivan.aliaga@urp.edu.pe>
// SPDX-License-Identifier: Apache-2.0
//
// Typed Dazzle-backed store, mirrors `ContextStore<T>` in
// Kotlin/Swift/Dart. Persists encodable values under a namespaced
// hash key (`<name>:<id>`).

import { dazzleCommand, RespArray, RespBulk } from '../ffi/command';
import { Dazzle } from '../primitives/dazzle';

export interface ContextStoreOptions<T> {
  name: string;
  encode: (value: T) => Record<string, string>;
  decode: (fields: Record<string, string>) => T | null;
}

export class ContextStore<T> {
  private readonly client = new Dazzle();

  constructor(private readonly opts: ContextStoreOptions<T>) {}

  private keyFor(id: string): string { return `${this.opts.name}:${id}`; }

  async put(id: string, value: T): Promise<void> {
    const fields = { ...this.opts.encode(value), __id: id };
    await this.client.hash(this.keyFor(id)).setAll(fields);
  }

  async get(id: string): Promise<T | null> {
    const fields = await this.client.hash(this.keyFor(id)).getAllDirect();
    if (!Object.keys(fields).length) return null;
    return this.opts.decode(fields);
  }

  async delete(id: string): Promise<boolean> {
    return this.client.hash(this.keyFor(id)).deleteKey();
  }

  /** Iterate every record in the store via SCAN. */
  async *iterate(pageSize = 256): AsyncIterable<[string, T]> {
    let cursor = '0';
    do {
      const r = await dazzleCommand(
          ['SCAN', cursor, 'MATCH', `${this.opts.name}:*`, 'COUNT', String(pageSize)]);
      if (!(r instanceof RespArray) || r.items.length < 2) break;
      const next = r.items[0] instanceof RespBulk ? r.items[0].value : '0';
      const keys: string[] = [];
      if (r.items[1] instanceof RespArray) {
        for (const e of r.items[1].items) {
          if (e instanceof RespBulk) keys.push(e.value);
        }
      }
      for (const k of keys) {
        const id = k.substring(this.opts.name.length + 1);
        const v = await this.get(id);
        if (v != null) yield [id, v];
      }
      cursor = next;
    } while (cursor !== '0');
  }
}

File: sdk/flutter/dazzle_flutter/lib/src/agent/context_store.dart
// Copyright 2026 Ivan Aliaga <ivan.aliaga@urp.edu.pe>
// SPDX-License-Identifier: Apache-2.0
//
// Typed Dazzle-backed store, mirroring `ContextStore<T>` in
// Kotlin/Swift. Persists any encodable value under a namespaced hash
// key ("<name>:<id>") using the snapshot-cache fast path.

import '../ffi/command.dart' show dazzleCommand, RespArray, RespBulk;
import '../primitives/dazzle.dart';

class ContextStore<T> {
  ContextStore({
    required this.name,
    required this.encode,
    required this.decode,
  });

  final String name;
  final Map<String, String> Function(T value) encode;
  final T? Function(Map<String, String> fields) decode;

  String _key(String id) => '$name:$id';

  /// Upsert a record. Overwrites field-by-field (hash merge semantics).
  void put(String id, T value) {
    final fields = Map<String, String>.from(encode(value))..['__id'] = id;
    Dazzle().hash(_key(id)).setAll(fields);
  }

  /// Fast-path read. Returns `null` if the hash doesn't exist or
  /// decoding fails.
  T? get(String id) {
    final fields = Dazzle().hash(_key(id)).getAllDirect();
    if (fields.isEmpty) return null;
    return decode(fields);
  }

  bool delete(String id) => Dazzle().hash(_key(id)).deleteKey();

  /// Iterate every record in the store. Uses cursor-based SCAN so big
  /// keyspaces don't block the main Isolate. Suitable for chat agents
  /// (<= a few hundred turns) or bulk restore.
  Iterable<(String, T)> iterate({int pageSize = 256}) sync* {
    var cursor = '0';
    while (true) {
      final reply = dazzleCommand(
          ['SCAN', cursor, 'MATCH', '$name:*', 'COUNT', '$pageSize']);
      if (reply is! RespArray || reply.items.length < 2) break;
      final nextCursor = switch (reply.items[0]) {
        RespBulk(:final value) => value,
        _ => '0',
      };
      final keys = switch (reply.items[1]) {
        RespArray(:final items) => items
            .whereType<RespBulk>()
            .map((e) => e.value)
            .toList(),
        _ => const <String>[],
      };
      for (final k in keys) {
        final id = k.substring(name.length + 1);
        final v = get(id);
        if (v != null) yield (id, v);
      }
      cursor = nextCursor;
      if (cursor == '0') break;
    }
  }
}
